django/db/models/functions/base.py

```diff
@@ -25,6 +25,9 @@ def __init__(self, expression, output_field):
     def as_sql(self, compiler, connection, **extra_context):
         if 'db_type' not in extra_context:
             extra_context['db_type'] = self.output_field.db_type(connection)
+        # Support CharField casting without max_length.
```

Comment:
Did you consider overriding `CharField.db_type`? There's [a related ticket](https://code.djangoproject.com/ticket/14094) proposing to allow using `CharField()` with `max_length` in models. This might partially allow this, although the system check is still there.